src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
             return new PointerTypeSymbol(TypeWithAnnotations.Create(elementType, elementNullableAnnotation));
         }
 
+        internal FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbol(
+            TypeWithAnnotations returnType,
+            ImmutableArray<TypeWithAnnotations> parameterTypes,
+            RefKind returnRefKind,
+            ImmutableArray<RefKind> parameterRefKinds)
+        {
+            if (returnType.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(returnType));
+            }
+
+            if (parameterTypes.IsDefault || parameterTypes.Any(param => param.IsDefault))
+            {
+                throw new ArgumentNullException(nameof(parameterTypes));
+            }
+
+            if (parameterRefKinds.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(parameterRefKinds));
+            }
+
+            if (parameterRefKinds.Length != parameterTypes.Length)
+            {
+                // Given {0} parameter types and {1} parameter ref kinds. These must be the same.
+                throw new ArgumentException(string.Format(CSharpResources.NotSameNumberParameterTypesAndRefKinds, parameterTypes.Length, parameterRefKinds.Length));
+            }
+
+            if (returnRefKind == RefKind.Out)
```

Comment:
I'm unsure what level of validation we should be doing here. For example, should we check that `void` isn't passed for a parameter type, and that `void` returns always have `RefKind.None`?

Reply:
> I'm unsure what level of validation we should be doing here. For example, should we check that void isn't passed for a parameter type, and that void returns always have RefKind.None?

I do not think we do this kind of validation in other similar scenarios, like constructing a generic type, anonymous type, etc.

---
In reply to: [428297776](https://github.com/dotnet/roslyn/pull/44436#discussion_r428297776) [](ancestors = 428297776)

Reply:
That said, I think it makes sense to keep the check for ```returnRefKind```.

---
In reply to: [428846682](https://github.com/dotnet/roslyn/pull/44436#discussion_r428846682) [](ancestors = 428846682,428297776)

Reply:
Read my mind about my next follow up question, thanks!

---
In reply to: [428848084](https://github.com/dotnet/roslyn/pull/44436#discussion_r428848084) [](ancestors = 428848084,428846682,428297776)